src/linear_algebra/quadratic_form.lean

```diff
@@ -104,31 +134,61 @@ by { cases Q, cases Q', congr, funext, apply H }
-    polar_smul_right := λ a x y, by simp [polar] } ⟩
+    polar_add_left' := λ x x' y, by simp [polar],
+    polar_smul_left' := λ a x y, by simp [polar],
+    polar_add_right' := λ x y y', by simp [polar],
+    polar_smul_right' := λ a x y, by simp [polar] } ⟩
+
+@[simp] lemma zero_apply (x : M) : (0 : quadratic_form R M) x = 0 := rfl
 
 instance : inhabited (quadratic_form R M) := ⟨0⟩
 
+instance : has_add (quadratic_form R M) :=
+⟨ λ Q Q',
+  { to_fun := Q + Q',
+    to_fun_smul := λ a x,
+      by simp only [pi.add_apply, map_smul, mul_add],
+    polar_add_left' := λ x x' y,
+      by simp only [polar_add, polar_add_left, add_assoc, add_left_comm],
+    polar_smul_left' := λ a x y,
+      by simp only [polar_add, smul_eq_mul, mul_add, polar_smul_left],
+    polar_add_right' := λ x y y',
+      by simp only [polar_add, polar_add_right, add_assoc, add_left_comm],
+    polar_smul_right' := λ a x y,
+      by simp only [polar_add, smul_eq_mul, mul_add, polar_smul_right] } ⟩
+
+@[simp] lemma add_apply (Q Q' : quadratic_form R M) (x : M) : (Q + Q') x = Q x + Q' x := rfl
+
 instance : has_scalar R₁ (quadratic_form R₁ M) :=
 ⟨ λ a Q,
   { to_fun := λ x, Q (a • x),
     to_fun_smul := λ b x, by rw [smul_comm, map_smul],
-    polar_add_left := λ x x' y,
-      by convert Q.polar_add_left (a • x) (a • x') (a • y) using 1; simp [polar, smul_add],
-    polar_smul_left := λ b x y,
-      by convert Q.polar_smul_left b (a • x) (a • y) using 1; simp [polar, smul_add, smul_comm],
-    polar_add_right := λ x y y',
-      by convert Q.polar_add_right (a • x) (a • y) (a • y') using 1; simp [polar, smul_add],
-    polar_smul_right := λ b x y,
-      by convert Q.polar_smul_right b (a • x) (a • y) using 1; simp [polar, smul_add, smul_comm] } ⟩
+    polar_add_left' := λ x x' y,
+      by convert polar_add_left (a • x) (a • x') (a • y) using 1; simp only [polar, smul_add],
+    polar_smul_left' := λ b x y,
+      by convert polar_smul_left b (a • x) (a • y) using 1;
+      simp only [polar, smul_add, smul_comm, smul_eq_mul],
+    polar_add_right' := λ x y y',
+      by convert polar_add_right (a • x) (a • y) (a • y') using 1; simp only [polar, smul_add],
+    polar_smul_right' := λ b x y,
+      by convert polar_smul_right b (a • x) (a • y) using 1;
+      simp only [polar, smul_add, smul_comm, smul_eq_mul] } ⟩
 
 @[simp] lemma smul_apply (a : R₁) (Q : quadratic_form R₁ M) (x : M) : (a • Q) x = Q (a • x) := rfl
 
-instance : mul_action R₁ (quadratic_form R₁ M) :=
+instance : add_comm_monoid (quadratic_form R M) :=
+{ add_comm := λ Q Q', by { ext, simp only [add_apply, add_comm] },
+  add_assoc := λ Q Q' Q'', by { ext, simp only [add_apply, add_assoc] },
+  add_zero := λ Q, by { ext, simp only [zero_apply, add_apply, add_zero] },
+  zero_add := λ Q, by { ext, simp only [zero_apply, add_apply, zero_add] },
+  ..quadratic_form.has_add,
+  ..quadratic_form.has_zero }
+
+instance : distrib_mul_action R₁ (quadratic_form R₁ M) :=
```

Comment:
Don't you even get a full blown module structure?

Reply:
We should, but I defined the wrong `has_scalar` instance before :P